registry/storage/garbagecollect.go

```diff
@@ -47,6 +49,17 @@ func MarkAndSweep(ctx context.Context, storageDriver driver.StorageDriver, regis
 		}
 
 		err = manifestEnumerator.Enumerate(ctx, func(dgst digest.Digest) error {
+			if manifestswithouttags {
+				tags, errr := repository.Tags(ctx).Lookup(ctx, distribution.Descriptor{Digest: dgst})
+				if errr != nil {
+					return fmt.Errorf("failed to retrieve tags for digest %v: %v", dgst, errr)
+				}
+				if len(tags) == 0 {
+					emit("manifest has no tags %s@%s", repoName, dgst)
+					manifestSet[dgst] = repoName
```

Comment:
@zetaab The file [`paths.go`](https://github.com/docker/distribution/blob/master/registry/storage/paths.go) documents the storage layout, including the relationships that must be maintained when modifying the storage layout. Any assumptions should be checked against the layout there.

I _think_ you might be able to get away with issuing the delete right here, since the reference set is contained in the tags present here and nowhere else. Effectively, this list of tags is acting like a local reference count.